File: apps/desktop/src/stores/shareControl.ts
/**
 * Giving somebody the mouse on a screen you are sharing in a call.
 *
 * This is not a remote-desktop session and it deliberately does not become one.
 * A remote session is a standing arrangement: the machine is enrolled, a grant
 * was written down beforehand, and a gateway enforces it. What happens in a
 * call is the opposite - two people are already talking, one of them is showing
 * the other a problem, and "here, you do it" should not need an administrator.
 * Teams calls it giving control, and the shape is the same one.
 *
 * So the whole exchange rides the peers' own data channels, directly between
 * the two clients, and the only authority is the person sharing clicking yes.
 * That is the right authority: it is their machine, they are sitting at it, and
 * they can see what is being done with it. Nothing here can start without them,
 * and one click ends it.
 *
 * What keeps it honest:
 *
 * - A data channel has exactly two ends. A message on the connection to a peer
 *   came from that peer and from nobody else - there is no server in between
 *   that could have written it, and the peer's identity came from the roster
 *   `call-service` built from authenticated sockets.
 * - Input is applied only from the one identity control was granted to, only
 *   while a screen is still being shared, and only when that share is a whole
 *   display - a window can be dragged between monitors, so there is no fraction
 *   of a screen to map a click onto, and control of one is refused.
 * - Control ends when the share ends, when the call does, when that person
 *   leaves, when either side presses the button, and when the person driving
 *   presses Escape.
 *
 * Pointers are the other half. Several people watch a share and any of them may
 * be pointing at something; each sends where their cursor is over the picture,
 * everyone draws everyone else's with a name on it, and a pointer that goes
 * quiet disappears. Only one person can drive, but everybody can point.
 */
import { create } from 'zustand';
import type { CallPeer } from '@nexora/shared-types';
import type { Mesh } from '../services/mesh';
// Circular by design and safe: the voice store attaches this one, and this one
// only reads it from inside a function, long after both modules have loaded.
import { useVoiceStore } from './voice';

/**
 * Mouse moves are sampled; the far end does not need 500 events a second.
 *
 * ponytail: one reliable, ordered data channel carries pointers, mouse moves
 * and grants alike, where the SFU build could mark the cosmetic ones lossy. At
 * these rates it does not matter; a second unordered channel is the fix if a
 * bad link ever makes pointers lag behind the picture.
 */
const MOVE_INTERVAL_MS = 25;
/** Pointers are cosmetic, so they are sent less often. */
const POINTER_INTERVAL_MS = 60;
/** A pointer nobody has heard from in this long has left the picture. */
const POINTER_STALE_MS = 4_000;

const TOPIC = 'nexora.share';

interface Person {
  identity: string;
  name: string;
}

/** A cursor to draw, in fractions of the shared screen. */
export interface SharePointer {
  identity: string;
  name: string;
  x: number;
  y: number;
  at: number;
}

type Wire =
  | { k: 'ask' }
  | { k: 'grant' }
  | { k: 'deny'; why: string }
  | { k: 'revoke' }
  | {
      k: 'm';
      a: 'move' | 'down' | 'up' | 'wheel';
      x: number;
      y: number;
      b?: 'left' | 'right' | 'middle';
      d?: number;
    }
  | { k: 'key'; a: 'down' | 'up'; key: string; code: string; mods?: string[] }
  | { k: 'p'; x: number; y: number }
  | { k: 'p.off' };

/** What actually crosses the channel, so a stray message is easy to ignore. */
interface Envelope {
  topic: string;
  message: Wire;
}

interface ShareControlState {
  /** Sharer side: people who have asked and have not been answered. */
  requests: Person[];
  /** Sharer side: who is driving this machine right now. */
  controller: Person | null;
  /** Viewer side: waiting for an answer. */
  asking: boolean;
  /** Viewer side: the identity whose screen this client is driving. */
  driving: string | null;
  /** Viewer side: why the last ask was turned down. */
  refusal: string | null;
  /** Everyone: identity -> where their cursor is over the share. */
  pointers: Record<string, SharePointer>;

  /** Called by the voice store when a call comes and goes. */
  attach: (mesh: Mesh) => void;
  detach: () => void;
  /** Called by the voice store for every data-channel message. */
  receive: (from: CallPeer, payload: unknown) => void;
  /** Called by the voice store when the roster changes. */
  peersChanged: (peers: CallPeer[]) => void;

  ask: (sharer: Person) => void;
  answer: (identity: string, granted: boolean) => void;
  /** Sharer takes it back, or the driver hands it over - same wire, same end. */
  stop: () => void;

  sendPointer: (x: number, y: number) => void;
  clearPointer: () => void;
  sendMouse: (
    action: 'move' | 'down' | 'up' | 'wheel',
    x: number,
    y: number,
    button?: 'left' | 'right' | 'middle',
    deltaY?: number,
  ) => void;
  /** `modifiers` is what was held at the moment of the press - see keyboard.ts. */
  sendKey: (action: 'down' | 'up', key: string, code: string, modifiers?: string[]) => void;
}

let mesh: Mesh | null = null;
let lastMoveAt = 0;
let lastPointerAt = 0;
let sweeper: number | null = null;
/** Who this client asked, so an unsolicited "yes" from anyone else is ignored. */
let asked: string | null = null;

function publish(message: Wire, to?: string[]): void {
  mesh?.sendData({ topic: TOPIC, message } satisfies Envelope, to);
}

export const useShareControlStore = create<ShareControlState>((set, get) => ({
  requests: [],
  controller: null,
  asking: false,
  driving: null,
  refusal: null,
  pointers: {},

  attach: (next) => {
    get().detach();
    mesh = next;

    // A cursor that stopped arriving has left the picture; nothing sends a
    // farewell when a window closes.
    sweeper = window.setInterval(() => {
      const cutoff = Date.now() - POINTER_STALE_MS;
      const pointers = Object.fromEntries(
        Object.entries(get().pointers).filter(([, pointer]) => pointer.at > cutoff),
      );
      if (Object.keys(pointers).length !== Object.keys(get().pointers).length) set({ pointers });
    }, 1_000);
  },

  detach: () => {
    if (sweeper !== null) window.clearInterval(sweeper);
    sweeper = null;
    mesh = null;
    asked = null;
    window.nexora?.remoteTarget(null, 'call');
    set({
      requests: [],
      controller: null,
      asking: false,
      driving: null,
      refusal: null,
      pointers: {},
    });
  },

  receive: (from, payload) => {
    const envelope = payload as Envelope | null;
    if (!envelope || envelope.topic !== TOPIC || !envelope.message) return;
    onMessage(envelope.message, from, set, get);
  },

  // Somebody who leaves cannot still be driving, and their cursor should not be
  // left on screen.
  peersChanged: (peers) => {
    const present = new Set(peers.map((peer) => peer.peerId));

    set((state) => {
      const pointers = Object.fromEntries(
        Object.entries(state.pointers).filter(([identity]) => present.has(identity)),
      );
      const controllerGone = state.controller !== null && !present.has(state.controller.identity);
      const drivenGone = state.driving !== null && !present.has(state.driving);

      return {
        pointers,
        requests: state.requests.filter((person) => present.has(person.identity)),
        controller: controllerGone ? null : state.controller,
        driving: drivenGone ? null : state.driving,
        asking: drivenGone ? false : state.asking,
      };
    });

    if (get().controller === null) window.nexora?.remoteTarget(null, 'call');
  },

  ask: (sharer) => {
    asked = sharer.identity;
    set({ asking: true, refusal: null, driving: null });
    publish({ k: 'ask' }, [sharer.identity]);
  },

  answer: (identity, granted) => {
    const request = get().requests.find((person) => person.identity === identity);
    set((state) => ({ requests: state.requests.filter((person) => person.identity !== identity) }));
    if (!request) return;

    if (!granted) {
      publish({ k: 'deny', why: 'declined' }, [identity]);
      return;
    }

    // The share can have stopped between the ask and the answer, and granting
    // control of a screen nobody is sending would be a grant with no picture
    // over it - the one situation where somebody is driving blind.
    const refusal = whyControlIsImpossible();
    if (refusal) {
      publish({ k: 'deny', why: refusal }, [identity]);
      return;
    }

    // Whatever was being controlled before is not any more: one driver.
    const previous = get().controller;
    if (previous && previous.identity !== identity) {
      publish({ k: 'revoke' }, [previous.identity]);
    }

    // Input arrives as a fraction of the shared display, so the main process
    // has to be told which display that is before the first event lands. Under
    // `call`, which is its own target: a machine can be in a remote session at
    // the same time, watching a different monitor, and the two must not
    // overwrite each other.
    window.nexora?.remoteTarget(sharedDisplayId(), 'call');
    set({ controller: request });
    publish({ k: 'grant' }, [identity]);
  },

  stop: () => {
    const { controller, driving } = get();
    if (controller) {
      publish({ k: 'revoke' }, [controller.identity]);
      window.nexora?.remoteTarget(null, 'call');
      set({ controller: null });
    }
    if (driving) {
      publish({ k: 'revoke' }, [driving]);
      asked = null;
      set({ driving: null, asking: false });
    }
  },

  sendPointer: (x, y) => {
    const now = Date.now();
    if (now - lastPointerAt < POINTER_INTERVAL_MS) return;
    lastPointerAt = now;
    publish({ k: 'p', x, y });
  },

  clearPointer: () => publish({ k: 'p.off' }),

  sendMouse: (action, x, y, button, deltaY) => {
    const target = get().driving;
    if (!target) return;
    if (action === 'move') {
      const now = Date.now();
      if (now - lastMoveAt < MOVE_INTERVAL_MS) return;
      lastMoveAt = now;
    }
    publish({ k: 'm', a: action, x, y, b: button, d: deltaY }, [target]);
  },

  sendKey: (action, key, code, modifiers) => {
    const target = get().driving;
    if (!target) return;
    publish({ k: 'key', a: action, key, code, mods: modifiers }, [target]);
  },
}));

type Setter = (
  partial:
    | Partial<ShareControlState>
    | ((state: ShareControlState) => Partial<ShareControlState>),
) => void;

function onMessage(
  message: Wire,
  from: CallPeer,
  set: Setter,
  get: () => ShareControlState,
): void {
  const identity = from.peerId;
  const name = from.username || from.peerId;

  switch (message.k) {
    case 'p':
      set((state) => ({
        pointers: {
          ...state.pointers,
          [identity]: { identity, name, x: message.x, y: message.y, at: Date.now() },
        },
      }));
      return;

    case 'p.off':
      set((state) => {
        const pointers = { ...state.pointers };
        delete pointers[identity];
        return { pointers };
      });
      return;

    // Somebody wants the mouse. Nothing is granted here - it goes in front of
    // the person sharing, and stays there until they answer.
    case 'ask': {
      const refusal = whyControlIsImpossible();
      if (refusal) {
        publish({ k: 'deny', why: refusal }, [identity]);
        return;
      }
      set((state) =>
        state.requests.some((person) => person.identity === identity)
          ? {}
          : { requests: [...state.requests, { identity, name }] },
      );
      return;
    }

    // Only from the person who was asked. An unsolicited "yes" is somebody
    // else's client being strange, and acting on it would point this window's
    // keyboard at a machine nobody offered.
    case 'grant':
      if (asked !== identity) return;
      asked = null;
      set({ driving: identity, asking: false, refusal: null });
      return;

    case 'deny':
      if (asked !== identity) return;
      asked = null;
      set({ asking: false, driving: null, refusal: message.why });
      return;

    // Sent by either side: the sharer taking it back, or the driver letting go.
    case 'revoke':
      if (get().controller?.identity === identity) {
        window.nexora?.remoteTarget(null, 'call');
        set({ controller: null });
      }
      if (get().driving === identity) set({ driving: null, asking: false });
      return;

    case 'm': {
      if (!mayDrive(identity, get)) return;
      window.nexora?.remoteMouse({
        action: message.a,
        x: message.x,
        y: message.y,
        button: message.b,
        deltaY: message.d,
        source: 'call',
      });
      return;
    }

    case 'key': {
      if (!mayDrive(identity, get)) return;
      window.nexora?.remoteKey({
        action: message.a,
        key: message.key,
        code: message.code,
        modifiers: message.mods,
        source: 'call',
      });
      return;
    }

    default:
      return;
  }
}

/**
 * The check that matters, applied to every single input event rather than once
 * at the grant: the sender has to be the one person control was given to, and
 * the screen has to still be being shared. A grant that outlived its share
 * would be a way to keep driving a machine nobody is watching.
 */
function mayDrive(identity: string, get: () => ShareControlState): boolean {
  return get().controller?.identity === identity && sharedDisplayId() !== null;
}

/**
 * Which display this client is sharing, or null when it is sharing a window or
 * nothing at all.
 */
function sharedDisplayId(): string | null {
  const { screenEnabled, sharedDisplayId: displayId } = useVoiceStore.getState();
  return screenEnabled ? displayId : null;
}

function whyControlIsImpossible(): string | null {
  const { screenEnabled, sharedDisplayId: displayId } = useVoiceStore.getState();
  if (!screenEnabled) return 'they are not sharing a screen';
  if (!displayId) return 'a window is being shared, not a whole screen';
  if (window.nexora?.platform !== 'win32') return 'control is not supported on that machine';
  return null;
}

File: apps/android/app/src/main/java/com/aatech/betweenus/feature/voice/ShareControl.kt
package com.aatech.betweenus.feature.voice

import androidx.compose.foundation.background
import androidx.compose.foundation.gestures.detectDragGestures
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import org.json.JSONObject

/**
 * Asking for the mouse on somebody's shared screen, from a phone.
 *
 * The viewer half of `apps/desktop/src/stores/shareControl.ts`, and only that
 * half. A phone can ask to drive and then drive; it cannot be driven, because
 * being driven means injecting synthetic mouse and keyboard events into the
 * machine, and Android has no such thing outside an accessibility service - a
 * permission this app has no business asking for.
 *
 * That asymmetry is not silent. An `ask` arriving here is answered with a
 * `deny` and a reason, so somebody on a desktop who asks a phone to hand over
 * its screen is told why not, rather than waiting on a dialog that will never
 * appear.
 *
 * ## Why this needs no server at all
 *
 * The whole exchange rides the peers' own data channels. A data channel has
 * exactly two ends: a message on the connection to a peer came from that peer
 * and from nobody else, and that peer's identity came from the roster
 * `call-service` built out of authenticated sockets. The only authority is the
 * person sharing pressing yes - which is the right authority, because it is
 * their machine and they are sitting at it.
 */
object ShareControl {

    /** The topic the desktop stamps these with. Must match, byte for byte. */
    const val TOPIC = "betweenus.share"

    /** Mouse moves are sampled; the far end does not need five hundred a second. */
    private const val MOVE_INTERVAL_MS = 25L

    /** Pointers are cosmetic, so they go less often. */
    private const val POINTER_INTERVAL_MS = 60L

    /** Waiting on an answer from this peer, or null. */
    private val _asking = MutableStateFlow<String?>(null)
    val asking: StateFlow<String?> = _asking.asStateFlow()

    /** The peer whose screen this phone is driving, or null. */
    private val _driving = MutableStateFlow<String?>(null)
    val driving: StateFlow<String?> = _driving.asStateFlow()

    /** Why the last ask was turned down, for the line that says so. */
    private val _refusal = MutableStateFlow<String?>(null)
    val refusal: StateFlow<String?> = _refusal.asStateFlow()

    /**
     * Who this client asked.
     *
     * Kept so an unsolicited "yes" from somebody who was never asked is
     * ignored: a grant is only meaningful from the peer the request went to.
     */
    private var asked: String? = null

    private var lastMoveAt = 0L
    private var lastPointerAt = 0L

    /** Sends one envelope to one peer. Set by [VoiceEngine] while a call runs. */
    internal var send: ((peerId: String, JSONObject) -> Unit)? = null

    fun attach(sender: (peerId: String, JSONObject) -> Unit) {
        send = sender
    }

    /** The call ended. Nothing is being driven, and nothing is waiting. */
    fun detach() {
        send = null
        asked = null
        _asking.value = null
        _driving.value = null
        _refusal.value = null
    }

    /** May I have the mouse? */
    fun ask(peerId: String) {
        asked = peerId
        _asking.value = peerId
        _refusal.value = null
        publish(peerId, JSONObject().put("k", "ask"))
    }

    /**
     * Hand it back.
     *
     * The same message the sharer sends to take it away: one wire for both
     * directions, because "this arrangement is over" is one fact whichever end
     * decided it.
     */
    fun release() {
        val peer = _driving.value ?: _asking.value ?: return
        publish(peer, JSONObject().put("k", "revoke"))
        asked = null
        _asking.value = null
        _driving.value = null
    }

    /**
     * One message from a peer's data channel.
     *
     * [fromPeerId] is who it came from, decided by which connection carried it
     * rather than by anything inside the message - a peer cannot claim to be
     * somebody else, because there is no third party in the middle to be
     * fooled.
     */
    internal fun receive(fromPeerId: String, message: JSONObject) {
        when (message.optString("k")) {
            // Somebody wants to drive *this* device. It has no mouse to give.
            "ask" -> publish(
                fromPeerId,
                JSONObject()
                    .put("k", "deny")
                    .put("why", "They are on a phone, which cannot be driven"),
            )

            "grant" -> {
                if (asked != fromPeerId) return
                asked = null
                _driving.value = fromPeerId
                _asking.value = null
                _refusal.value = null
            }

            "deny" -> {
                if (asked != fromPeerId) return
                asked = null
                _asking.value = null
                _driving.value = null
                _refusal.value = message.optString("why").ifBlank { "They said no" }
            }

            // Either side ending it: the sharer taking it back, or an echo of
            // this client letting go.
            "revoke" -> {
                if (_driving.value == fromPeerId) _driving.value = null
                if (_asking.value == fromPeerId) {
                    asked = null
                    _asking.value = null
                }
            }

            // Input aimed at this device. There is nothing here to inject it
            // into, and dropping it is the whole of the correct behaviour.
            else -> Unit
        }
    }

    /**
     * A tap or a drag on the shared picture, in fractions of it.
     *
     * Fractions rather than pixels because only the machine at the far end
     * knows what its screen measures - the same reason the remote-desktop
     * viewer sends fractions. A phone showing a 4K desktop letterboxed into a
     * tile has no pixel of its own worth naming.
     *
     * Moves are throttled; presses and releases never are. A dropped move is a
     * slightly coarser path, and a dropped press is a click that did not
     * happen.
     */
    fun sendMouse(
        action: String,
        x: Float,
        y: Float,
        button: String? = null,
        deltaY: Float? = null,
    ) {
        val peer = _driving.value ?: return
        if (action == "move") {
            val now = System.currentTimeMillis()
            if (now - lastMoveAt < MOVE_INTERVAL_MS) return
            lastMoveAt = now
        }
        publish(
            peer,
            JSONObject()
                .put("k", "m")
                .put("a", action)
                .put("x", x.coerceIn(0f, 1f).toDouble())
                .put("y", y.coerceIn(0f, 1f).toDouble())
                .apply {
                    button?.let { put("b", it) }
                    deltaY?.let { put("d", it.toDouble()) }
                },
        )
    }

    /**
     * Where this person is pointing, for everybody watching.
     *
     * Only one person can drive; anybody can point. Sent to the sharer, who is
     * the one client that can place it on the picture everybody else is
     * watching a copy of.
     */
    fun sendPointer(sharerPeerId: String, x: Float, y: Float) {
        val now = System.currentTimeMillis()
        if (now - lastPointerAt < POINTER_INTERVAL_MS) return
        lastPointerAt = now
        publish(
            sharerPeerId,
            JSONObject()
                .put("k", "p")
                .put("x", x.coerceIn(0f, 1f).toDouble())
                .put("y", y.coerceIn(0f, 1f).toDouble()),
        )
    }

    fun clearPointer(sharerPeerId: String) =
        publish(sharerPeerId, JSONObject().put("k", "p.off"))

    private fun publish(peerId: String, message: JSONObject) {
        send?.invoke(peerId, JSONObject().put("topic", TOPIC).put("message", message))
    }
}

// --- the screen half -----------------------------------------------

/**
 * The bar that asks, waits, and says what came of it.
 *
 * Drawn under the share itself, which is the only place on a phone where it is
 * about something the person can see. It used to sit on the call stage, above
 * the dock - where it was unreachable in practice, because a share took the
 * whole screen the instant it started and this bar was on the screen it
 * replaced.
 *
 * The button is deliberately *not* [com.aatech.betweenus.ui.components.BetweenUsButton]:
 * that one fills its width, which in a `Row` measures before the weighted text
 * and left the label with nothing. A bar whose caption is one ellipsis is the
 * "buttons are not properly raised" bug.
 */
@Composable
fun ShareControlBar(
    sharerPeerId: String,
    sharerName: String,
    modifier: Modifier = Modifier,
) {
    val scheme = MaterialTheme.colorScheme
    val asking by ShareControl.asking.collectAsStateWithLifecycle()
    val driving by ShareControl.driving.collectAsStateWithLifecycle()
    val refusal by ShareControl.refusal.collectAsStateWithLifecycle()

    val amDriving = driving == sharerPeerId
    val amAsking = asking == sharerPeerId

    Row(
        modifier = modifier
            .fillMaxWidth()
            .clip(MaterialTheme.shapes.large)
            .background(
                if (amDriving) scheme.primaryContainer else scheme.surfaceContainerHigh,
            )
            .padding(start = 14.dp, end = 8.dp, top = 8.dp, bottom = 8.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        Text(
            text = when {
                // Said here rather than discovered by pinching and having
                // nothing happen: while this phone is the mouse, a drag is a
                // drag on their machine and cannot also be a zoom on this one.
                amDriving -> "Driving $sharerName's screen - tap and drag to use it"
                amAsking -> "Asked $sharerName for the mouse…"
                refusal != null -> refusal.orEmpty()
                else -> "Ask $sharerName for the mouse to use their screen"
            },
            style = MaterialTheme.typography.bodySmall,
            color = if (amDriving) scheme.onPrimaryContainer else scheme.onSurfaceVariant,
            maxLines = 2,
            overflow = TextOverflow.Ellipsis,
            // `weight` and nothing else: the button sizes to its own words and
            // the caption takes what is left, which is the way round that keeps
            // both readable on a 320dp phone.
            modifier = Modifier.weight(1f),
        )

        Button(
            onClick = {
                when {
                    amDriving || amAsking -> ShareControl.release()
                    else -> ShareControl.ask(sharerPeerId)
                }
            },
            shapes = ButtonDefaults.shapes(),
            contentPadding = ButtonDefaults.ContentPadding,
            modifier = Modifier.heightIn(min = 40.dp),
        ) {
            Text(
                text = when {
                    amDriving -> "Stop"
                    amAsking -> "Cancel"
                    else -> "Ask to drive"
                },
                style = MaterialTheme.typography.labelLarge,
                maxLines = 1,
            )
        }
    }
}

/**
 * The picture, while this phone is driving it.
 *
 * Touches become fractions of this box, so the box has to *be* the picture -
 * see the call site in [ShareStage], which sizes it to the letterboxed frame
 * rather than to the black around it. A drag is a press, a run of moves and a
 * release, which is what a mouse does and what the far end injects.
 *
 * Not a trackpad. A phone driving a desktop is somebody pointing at a thing on
 * a screen they can see, and an absolute surface is the one that matches what
 * they are looking at; a relative pad would need an on-screen cursor this app
 * would then have to draw and keep in step with the far end's own.
 */
@Composable
fun DriveSurface(sharerPeerId: String, modifier: Modifier = Modifier) {
    val driving by ShareControl.driving.collectAsStateWithLifecycle()
    if (driving != sharerPeerId) return

    Box(
        modifier = modifier.pointerInput(sharerPeerId) {
            detectDragGestures(
                onDragStart = { at ->
                    ShareControl.sendMouse("move", at.x / size.width, at.y / size.height)
                    ShareControl.sendMouse(
                        "down",
                        at.x / size.width,
                        at.y / size.height,
                        button = "left",
                    )
                },
                onDrag = { change, _ ->
                    ShareControl.sendMouse(
                        "move",
                        change.position.x / size.width,
                        change.position.y / size.height,
                    )
                },
                onDragEnd = { ShareControl.sendMouse("up", 0f, 0f, button = "left") },
                // A cancelled gesture must still release the button. A press
                // that is never lifted is a mouse held down on somebody else's
                // machine, which is the worst thing this can leave behind.
                onDragCancel = { ShareControl.sendMouse("up", 0f, 0f, button = "left") },
            )
        },
    ) {
        // A tap is a click in place: press and release at the same point.
        Box(
            Modifier
                .matchParentSize()
                .pointerInput(sharerPeerId) {
                    detectTapGestures { at ->
                        val x = at.x / size.width
                        val y = at.y / size.height
                        ShareControl.sendMouse("move", x, y)
                        ShareControl.sendMouse("down", x, y, button = "left")
                        ShareControl.sendMouse("up", x, y, button = "left")
                    }
                },
        )
    }
}
